### ISSUE TITLE
[DONE; update documentation] publish Polly v6.0.0 as strong-named only

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#442]
### Background

ASP.NET Core 2.1 is moving to consume Polly for `HttpClient` instances created through [`HttpClientFactory`](https://github.com/aspnet/HttpClientFactory). This will allow the easy application of Polly resilience strategies to outbound calls through `HttpClient`.  

This will significantly increase consumption of Polly's strong-named package (currently labelled [Polly-Signed](https://www.nuget.org/packages/polly-Signed)).  That however will [create conflicts](https://github.com/BrighterCommand/Brighter/issues/286) for packages or projects that currently consume non-strong-named Polly: [example](https://github.com/BrighterCommand/Brighter/issues/286); [longer discussion](https://github.com/aspnet/HttpClientFactory/issues/105).

_There is no easy answer to this problem_, as attested by discussion on _numerous_ projects: [octokit](https://github.com/octokit/octokit.net/issues/405) and the _many_ projects that reference that issue.  

Options:

+ ~**Be purely non-strong-named**~: project cannot then be consumed in environments or by upstream packages mandating strong-naming; not an option for Polly due to ASP.NET Core.  
+ **Offer both strong-named and not** (as Polly currently): but [creates conflicts](https://github.com/BrighterCommand/Brighter/issues/286) where a project ends up transiently referencing both.

Or:

## Publish only strong-named

Some consensus among high-consumption projects now is around this, the [NewtonSoft.Json strategy](https://github.com/aspnet/HttpClientFactory/issues/105#issuecomment-385527782), as the least of all evils. @jeremyskinner of FluentValidation [sums it up well here](https://github.com/JeremySkinner/FluentValidation/issues/482#issuecomment-323944469) as does @sharwell [elsewhere](https://github.com/restsharp/RestSharp/issues/216#issuecomment-39761855).  I summarise also [here](https://github.com/aspnet/HttpClientFactory/issues/105#issuecomment-385527782).

**TL;DR** Switching to release Polly as strong-named only would be a _one-time hit_ breaking change for consumers of Polly.  _But thereafter_, having _only_ the strong-named version of Polly in the market would _prevent/reduce [those conflicts](https://github.com/BrighterCommand/Brighter/issues/286)_ (strong-named versus not) _going forward_.  

(edited: for typos)

---

_This is a **proposal** on which we would appreciate evidence-driven feedback._ Particularly any consumption scenarios we may not be aware of.  The proposed change is driven by the HttpClientFactory ASP.NET Core integration, which will drive many more Polly users to the strong-named version.  ASP.NET Core has prior decision to be strong-named and thus consume strong-named assemblies.

**Comment by joelhulen** [comment]
I am leaning toward publishing only strong-named, starting with version 6.0. This would result in us ceasing the `Polly-Signed` lineage, and we'd need to make it very clear in several places that the assembly in the `Polly` package is strong-named.

I had initially thought of having a new namespace called something like PollyCore that is strongly signed, and maintain that separately like we do with Polly and Polly-Signed, in order to reduce conflicts. Then thought better of it as adding more confusion and then dealing with ambiguous method names, etc. when you bring in `HttpClientFactory` to an existing project that uses Polly already. The other downside is that you'd have two identical DLLs, but with different names. This is not ideal for various reasons.

**Comment by JamesNK** [comment]
> I am leaning toward publishing only strong-named, starting with version 6.0. This would result in us ceasing the Polly-Signed lineage, and we'd need to make it very clear in several places that the assembly in the Polly package is strong-named.

On .NET Framework any library that depends on the package will break on upgrade until it is recompiled. That is fine for apps for other NuGet packages depending on it will break until upgraded.

**Comment by iancooper** [comment]
I commented over [here](https://github.com/aspnet/HttpClientFactory/issues/105) that consistency in the ecosystem is probably the best outcome we can hope for. The push to signed policy created by asp.net HttpClientFactory is essentially a fait accompli here, we are all going to have to take the signed version. So it would be easier to create a pit of success.

Strong naming is problematic, but I won't solve that problem...

**Comment by reisenberger** [comment]
Briefly to add some stats around the dependent packages issue highlighted ^^ by @JamesNK (thanks!), I don't know if the following stats are accurate/complete, but [libraries.io](https://libraries.io) helpfully gives us some idea of the packages currently dependent on Polly unsigned:

+ [dependent packages](https://libraries.io/nuget/Polly/dependents) - 163
+ [repos](https://libraries.io/nuget/Polly/usage) - 267

**Comment by holytshirt** [comment]
Brighter project will depend on this change
https://github.com/BrighterCommand/Brighter/issues/287

**Comment by sharwell** [comment]
💭 Is adding a strong name a binary breaking change for applications that are not strong-named? It's been so long since I did anything with an assembly that wasn't strong named that I'm not sure the answer.

**Comment by reisenberger** [comment]
x-ref [input from the Dot Net team](https://github.com/aspnet/HttpClientFactory/issues/105#issuecomment-385700108)

**Comment by iancooper** [comment]
@sharwell Yes, see this Brighter issue where we are tracking: https://github.com/BrighterCommand/Brighter/issues/286

**Comment by iancooper** [comment]
But, we are happy to go this route. It seems to be the best path forward. We'll have to release as a 'breaking change' but we can live with that issue.

**Comment by reisenberger** [comment]
Thanks @iancooper and @rvanmaanen for highlighting this important issue.

x-ref https://github.com/aspnet/HttpClientFactory/issues/108 where I am querying the ASPNET Core team on the timing for this.

**Comment by reisenberger** [comment]
@pardahlman (for [RawRabbit](https://github.com/pardahlman/RawRabbit)); @TomPallister (for [Ocelot](https://github.com/ThreeMammals/Ocelot)); @fanliang11 (for [surging](https://github.com/dotnetcore/surging)): @RamjotSingh (for [Microsoft.Bot.Connector.Teams](https://github.com/OfficeDev/BotBuilder-MicrosoftTeams))

[libraries.io suggests](https://libraries.io/nuget/Polly/dependents) you maintain libraries which have high downloads on nuget or high github stars, and consume Polly as a dependency. As a courtesy, please read this thread: Polly intends, from Polly v6.0.0, to switch to publish only as a strong-named version.  

Polly v6.0.0 will be a release functionally identical to Polly v5.9.0, but a major-version bump (v6.0.0) as we switch to purely strong-naming.

### Current (up to Polly v5.9.0)

| nuget package | kind |
| --- | --- |
| [Polly](https://www.nuget.org/packages/Polly/) | non strong-named |
| [Polly-Signed](https://www.nuget.org/packages/Polly-Signed/) |  strong-named |

### Polly v6.0.0 onwards

| nuget package | kind |
| --- | --- |
| Polly | strong-named |

### Impact

*Both non-strong-named DLLs and strong-named DLLs can reference strong-named DLLs.*  

Whether the DLLs in your package are themselves strong-named or not, you will be able to reference the new Polly v6.0.0 strong-named package.

### Why is Polly doing this?

ASPNET Core 2.1 is moving to consume Polly for `HttpClient` instances created through [`HttpClientFactory`](https://github.com/aspnet/HttpClientFactory). This will allow the easy application of Polly resilience strategies to outbound calls through `HttpClient`.  

This is expected to significantly increase consumption of Polly's strong-named package (currently labelled [Polly-Signed](https://www.nuget.org/packages/polly-Signed)).  That however will [create conflicts _which apparently have no solution_](https://github.com/BrighterCommand/Brighter/issues/286) for packages or projects that currently consume non-strong-named Polly ([example](https://github.com/BrighterCommand/Brighter/issues/286)) ([good summary](https://github.com/JeremySkinner/FluentValidation/wiki/l.-Strong-naming-and-signed-packages)) and wish to use the new ASPNET Core 2.1 features.  

If we leave both non-strong-named and strong-named Polly in the market, we prolong the possibility of these conflicts into future months and years.  We prolong the possibility that we (and you) will have users who complain that products are unusable, because some product (perhaps consuming yours) requires strong-named Polly as a transitive dependency via one axis, and non-strong-named Polly as a transitive dependency via another axis.

If we switch to a single, strong-named Polly release, there is initial impact (the same conflicts) while products switch to strong-named Polly, but the _duration of pain is limited_ by there being only a single version in the market going forward.

Deeper discussion: [1](https://github.com/App-vNext/Polly/issues/442#issue-319048428); [2](https://github.com/aspnet/HttpClientFactory/issues/105).

This strategy is the strategy used by many high-download nuget packages such as NewtonSoft.Json, FluentValidation, and Serilog, and is the strategy [recommended by the Microsoft ASP.NET team](https://github.com/aspnet/HttpClientFactory/issues/105#issuecomment-385700108).

### Reducing onward impact

We are also adopting the so-called [Newtonsoft.Json strategy described here](https://github.com/aspnet/HttpClientFactory/issues/105#issuecomment-385527782) to minimise onward impact.  In this strategy:

+ nuget package numbers follow [full semver](https://semver.org) (Major.Minor.Patch increments) to differentiate bug-fixes, new-functionality-without-breaking-changes, and breaking-changes.
+ Dlls [_partially_]((https://github.com/aspnet/HttpClientFactory/issues/105#issuecomment-385527782)) follow Major.0.0 version numbering only (even if the true version is Major.Minor.Patch).  This reduces the burden on projects which consume Polly of introducing excessive assembly binding redirects.

---

Following community discussion, we believe this to be the best way forward (but comments and questions are welcome!).

**Comment by joelhulen** [comment]
We've just released [Polly version 6.0.0-alpha](https://www.nuget.org/packages/Polly/6.0.0-v6alpha0001) to nuget.org. It has the following characteristics:

    - Publish as strong-named package only (discontinue non-strong-named versions)
    - Add .NetStandard 2.0 tfm
    - Provide .NET4.5 support via .NetStandard 1.1 tfm
    - Discontinue .NET4.0 support
    - Remove methods marked as deprecated in v5.9.0

The first item is the one that addresses the primary issue. Because this is related to Polly and its relationship to ASP.NET Core 2.1,we're working on releasing Polly.Extensions.Http 2.0 alpha tomorrow or the next day, which follows the same single-strong-name package method and will reference this new Polly v6.0 package.

**Comment by TomPallister** [comment]
Thanks for the heads up, Ocelot uses Polly so I will do whatever I need to fit in with you :)

Thanks for the amazing work and the Polly project!

**Comment by reisenberger** [comment]
[Polly v6.0.1](https://www.nuget.org/packages/Polly/6.0.1) is published, delivering the changes discussed in this thread.

Leaving the issue open as I intend to transition the discussion from this thread and https://github.com/aspnet/HttpClientFactory/issues/108, https://github.com/aspnet/HttpClientFactory/issues/105 into the Polly wiki, as documentation.

**Comment by reisenberger** [comment]
Documentation [published to wiki](https://github.com/App-vNext/Polly/wiki/Polly-v6-breaking-changes)
